<a href="https://colab.research.google.com/github/worldstar0722/IS4490_FA26/blob/main/Choi_Ellie_Module2_LocalModelTaskPortfolio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

> ### Note on Labs and Assignments:
>
> 🔧 Look for the **wrench emoji** — it marks code you must change. Routine run-only cells do not use it.
>
> 🖊 Look for the **writing emoji** — it marks analysis you must write.
>
> These sections are graded and are not optional.
>

# Module 2 Assignment 2: Local Model Task Portfolio

**Notebook:** Student Template  
**Student:** Edit the configuration cell  
**Required models:** `gemma3:1b`, `gemma3:4b`, `llama3.2:1b`, and `llama3.2:3b`

This notebook supports the complete Assignment 2 workflow: four direct business
tasks, prompt revision, a controlled four-model comparison, evidence-based scoring, reflection, and AI-use disclosure.


## Student Introduction: What is this Assignment Is About?

This assignment asks you to work directly with small language models running on your own machine via Ollama. You will practice two core skills:

1. **Prompt engineering** — writing and iteratively improving instructions that guide a model toward a useful business output.
2. **Model evaluation** — systematically comparing four models on the same task and scoring them with evidence.

**What you will produce:**

- **Part 1:** Four business tasks (extraction, summarization, drafting, classification). For each task, you write an initial zero-shot instruction, run it, diagnose a specific weakness in the output, revise the instruction using a named prompting strategy, run it again, and evaluate the improvement.
- **Part 2:** A controlled four-model comparison using a shared service-request packet. You design a single instruction, run it on all four models without changing anything, then score each model across four dimensions with evidence from their outputs.
- **Part 3:** A 350–500 word reflection answering seven specific questions about what you observed.

**Before you start:**
1. Run the initial code blocks to install Ollama and start it.
2. Replace `"Your Name"` in the configuration cell below with your actual name.
3. Run all cells from top to bottom in order.

The notebook will raise an error and stop if any required `TODO` is still present when you try to run a model — this is intentional so you do not accidentally submit incomplete work.

## Important Instructions

1. Read the assignment before editing this notebook.
2. Edit only cells marked for student work.
3. Do not change the comparison source packet, model list, or shared settings.
4. Preserve the first output from every run.
5. Before submitting, restart the kernel and run all cells from top to bottom.

The template intentionally raises a clear error when a required `TODO` remains.


## Setup Ollama

This notebook will download, install and start [Ollama](https://ollama.com/download). The four default model downloads require approximately 8 GB in total.





### What is Ollama?

Ollama is a tool that lets you run AI language models on your own computer instead of only using an online service like ChatGPT.

For a beginner, you can think of it as a local “AI model manager.” It helps you download a model, start it, and send it prompts. For example, instead of calling an online API from OpenAI, Google, or Anthropic, you can call an Ollama model running on your laptop or server.

The basic idea is:



*   You install Ollama.
*   You download a model, such as Llama, Gemma, or Mistral.
*   You send text to the model.
*  The model sends text back.


Why are we doing this rather than using Claude or ChatGPT?

* Reproducability. The notebook allows students to all follow the same steps and instructions.
* Cost: API access to Anthropic,OpenAI, Google models is not free. These models are free to run. The trade-off? (There always is one).
* What do we sacrifice for using the free models? Quality of responses and speed. We'll be using CPUs since these models are small language models (SLMs) rather than large language models (LLMs)






In [1]:
import subprocess
import time

In [2]:
# Download and install Ollama (Google Colab only — skip if running locally)
install_zstd = subprocess.run(
    "sudo apt-get install zstd",
    shell=True,
    capture_output=True,
    text=True,
)

install_zstd

CompletedProcess(args='sudo apt-get install zstd', returncode=0, stdout='Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following NEW packages will be installed:\n  zstd\n0 upgraded, 1 newly installed, 0 to remove and 57 not upgraded.\nNeed to get 603 kB of archives.\nAfter this operation, 1,695 kB of additional disk space will be used.\nGet:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 zstd amd64 1.4.8+dfsg-3build1 [603 kB]\nFetched 603 kB in 1s (560 kB/s)\nSelecting previously unselected package zstd.\n(Reading database ... \n(Reading database ... 5%\n(Reading database ... 10%\n(Reading database ... 15%\n(Reading database ... 20%\n(Reading database ... 25%\n(Reading database ... 30%\n(Reading database ... 35%\n(Reading database ... 40%\n(Reading database ... 45%\n(Reading database ... 50%\n(Reading database ... 55%\n(Reading database ... 60%\n(Reading database ... 65%\n(Reading database ... 70%\n(Reading database ... 75%\n(Reading datab

In [3]:
# RUN THIS CELL. YOU SHOULD SEE Ollama installed and Ollama server is running messages.

# Download and install Ollama
install = subprocess.run(
    "curl -fsSL https://ollama.com/install.sh | sh",
    shell=True,
    capture_output=True,
    text=True,
)
if install.returncode != 0:
    raise RuntimeError(f"Ollama installation failed:\n{install.stderr}")
print("Ollama installed.")

# Start the Ollama server as a background process
subprocess.Popen(
    ["ollama", "serve"],
    stdout=subprocess.DEVNULL,
    stderr=subprocess.DEVNULL,
)

# Give the server a few seconds to initialize before any requests are made
time.sleep(3)
print("Ollama server is running.")

Ollama installed.
Ollama server is running.


In [4]:
# RUN THIS CELL

from datetime import datetime
from hashlib import sha256
from time import perf_counter
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen
import json
import textwrap

from IPython.display import Markdown, display
import subprocess
import time


REFERENCE_MODE = False
AUTO_PULL_MODELS = True
OLLAMA_BASE_URL = "http://localhost:11434"

REQUIRED_MODELS = [
    "gemma3:1b",
    "gemma3:4b",
    "llama3.2:1b",
    "llama3.2:3b",
]
BASELINE_MODEL = "gemma3:1b"
GENERATION_OPTIONS = {
    "temperature": 1,
    #"seed": 4490,
    "num_ctx": 8192,
    "num_predict": 900,
}




In [5]:
# RUN THIS CELL

REFERENCE_OUTPUTS = {}


def require_finished(label, value):
    """Stop before a model run when a required student field is unfinished."""
    if value is None or "TODO" in str(value):
        raise ValueError(f"Complete {label} before running this cell.")


def ollama_request(path, payload=None, timeout=120):
    """Send a JSON request to the local Ollama service."""
    data = None if payload is None else json.dumps(payload).encode("utf-8")
    request = Request(
        f"{OLLAMA_BASE_URL}{path}",
        data=data,
        headers={"Content-Type": "application/json"},
        method="GET" if payload is None else "POST",
    )
    try:
        with urlopen(request, timeout=timeout) as response:
            return json.loads(response.read().decode("utf-8"))
    except HTTPError as exc:
        details = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(
            f"Ollama returned HTTP {exc.code}: {details}"
        ) from exc
    except URLError as exc:
        raise RuntimeError(
            "Cannot connect to Ollama at http://localhost:11434. "
            "Install and start Ollama, then rerun this cell."
        ) from exc


def chat_once(model, prompt, run_key):
    """Run one independent prompt and return content plus observable metadata."""
    if REFERENCE_MODE:
        fixture = REFERENCE_OUTPUTS[run_key]
        return {
            "model": model,
            "run_key": run_key,
            "recorded_at": fixture["recorded_at"],
            "elapsed_seconds": fixture["elapsed_seconds"],
            "content": fixture["content"],
            "prompt_eval_count": None,
            "eval_count": None,
            "reference_fixture": True,
        }

    started_at = datetime.now().astimezone().isoformat(timespec="seconds")
    start = perf_counter()
    response = ollama_request(
        "/api/chat",
        {
            "model": model,
            "messages": [{"role": "user", "content": prompt}],
            "stream": False,
            "keep_alive": 0,
            "options": GENERATION_OPTIONS,
        },
        timeout=900,
    )

    #print(f"Generation Options:{GENERATION_OPTIONS}")

    elapsed = round(perf_counter() - start, 2)
    return {
        "model": model,
        "run_key": run_key,
        "recorded_at": started_at,
        "elapsed_seconds": elapsed,
        "content": response["message"]["content"].strip(),
        "prompt_eval_count": response.get("prompt_eval_count"),
        "eval_count": response.get("eval_count"),
        "reference_fixture": False,
    }


def display_record(record):
    metadata = (
        f"**Model:** `{record['model']}`  \n"
        f"**Recorded:** {record['recorded_at']}  \n"
        f"**Elapsed:** {record['elapsed_seconds']} seconds"
    )
    display(Markdown(metadata))
    display(Markdown(record["content"]))


def compose_prompt(instruction, source):
    return (
        instruction.strip()
        + "\n\nSOURCE\n------\n"
        + source.strip()
    )


def run_part1(task_key, stage, instruction, source):
    require_finished(f"{task_key} {stage} instruction", instruction)
    return chat_once(
        BASELINE_MODEL,
        compose_prompt(instruction, source),
        f"part1_{task_key}_{stage}",
    )


In [6]:
# RUN THIS CELL
# You should see
# Attempting to pull model: gemma3:1b
# Successfully pulled gemma3:1b.
# etc.

# Check and pull required models if AUTO_PULL_MODELS is True
if AUTO_PULL_MODELS:
    print(f"Checking and pulling required models: {REQUIRED_MODELS}")
    for model_name in REQUIRED_MODELS:
        print(f"Attempting to pull model: {model_name}")
        # Use subprocess to run ollama pull command
        pull_result = subprocess.run(
            f"ollama pull {model_name}",
            shell=True,
            capture_output=True,
            text=True,
        )
        if pull_result.returncode != 0:
            print(f"Failed to pull {model_name}:\n{pull_result.stderr}")
        else:
            print(f"Successfully pulled {model_name}.")
        time.sleep(1) # Give a moment between pulls

Checking and pulling required models: ['gemma3:1b', 'gemma3:4b', 'llama3.2:1b', 'llama3.2:3b']
Attempting to pull model: gemma3:1b
Successfully pulled gemma3:1b.
Attempting to pull model: gemma3:4b
Successfully pulled gemma3:4b.
Attempting to pull model: llama3.2:1b
Successfully pulled llama3.2:1b.
Attempting to pull model: llama3.2:3b
Successfully pulled llama3.2:3b.


# Part 1: Direct AI Task Portfolio

In this part you complete four independent business tasks using the baseline model (`gemma3:1b`). Each task gives you a realistic business scenario and a source text. Your job is to engineer the prompt that produces the most useful output.

**How each task section works:**

1. **Write a zero-shot instruction** in the first code cell. Zero-shot means plain directions only — no examples, no step-by-step reasoning prompts. The code comment already labels it for you.
2. **Run the cell and preserve the output.** Do not delete or re-run the initial output before recording it. Your submission must show the original output.
3. **Diagnose the output** in the markdown cell that follows. Compare what the model produced against the source text. Identify one specific, evidence-based weakness (something missing, wrong, or poorly formatted). Quote or paraphrase both the source and the output.
4. **Write a revised instruction** in the second code cell, addressing the weakness you identified. For **at least two of the four tasks**, apply a named prompting strategy and identify it at the top of your instruction string.
5. **Evaluate the revision** in the final markdown cell. Explain whether the revision materially improved the output, what role the chosen strategy played, and what decisions still require a human's judgment.

**Named strategies you may apply for revised instructions:**

| Strategy | What it means |
|---|---|
| **Few-shot** | Include one or more examples of the expected input/output pattern before the task. |
| **Chain-of-thought** | Instruct the model to reason step by step before giving its final answer. |
| **Persona** | Assign the model a specific role or professional background before the task. |
| **Zero-shot** | Plain directions only — acceptable when the initial output already meets your standards. |

Preserve every initial output before revising an instruction.

In [7]:
PART1_SOURCES = {
  "extraction": "From: Maya Chen\nTo: Facilities Service Desk\nSubject: Loose handrail before Friday tour\n\nThe handrail in the east stairwell on floor 3 of Pioneer Hall is loose at the\nlower wall bracket. I noticed it at 9:15 a.m. on September 2. No one has been\ninjured, and the stairwell is still open. Please repair it before the visitor\ntour this Friday if possible. I can meet a technician after 1:00 p.m. Call me\nat extension 5521.",
  "summarization": "Customer Elena Ruiz reported that order OR-8841 was charged twice. The\noriginal $186.40 charge posted on August 6, and a second $186.40 charge posted\non August 8 after she refreshed the checkout page. The order itself arrived on\nAugust 10 and was correct. Agent Malik opened case CS-2197 on August 11 and\nasked Billing to investigate. Billing has not yet confirmed whether the second\nentry is a settled charge or a temporary authorization. Elena wants the second\ncharge removed if it settled, but she does not want the order canceled. She\nasked for an update by August 13 because her card payment is due August 14.",
  "drafting": "Supplier Northstar Filtration notified Procurement that shipment NF-771,\ncontaining 12 replacement filters, will arrive August 19 instead of August 14.\nThe plant currently has approximately four days of filter inventory at normal\nusage. Northstar offered expedited shipping for an additional fee, but\nProcurement has not approved that option. Operations is checking whether usage\ncan be reduced safely. The plant manager needs a status update today. No\nproduction shutdown has been scheduled.",
  "classification": "Routing categories:\n- IT Support: computers, software, networks, and accounts\n- Security Access: badges, controlled doors, and physical-access permissions\n- Facilities: building fixtures, utilities, and room conditions\n\nUrgency rules:\n- Urgent: an active safety issue or current business operation is blocked with\n  no workaround\n- Standard: future need, routine repair, or a workable temporary alternative\n\nRequest: \"My new analyst starts Monday. Her employee account works, but her\nbadge does not open the Finance Annex. I can meet her in the lobby and escort\nher on the first day if needed. Please add normal weekday access before 8:00\na.m. Monday. The request does not include the analyst's employee ID or the\nmanager's access approval record.\" "
}


## Part 1.1: Extraction

In this section you are fulfilling the role of an AI Automation Specialist working as a consultant for a local facilities management operation. They have been struggling with the amount of time it takes to read and synthesize information from unstructured emails from customers at various facilities. Your job is to extract key items from emails.

Your first prompt must be a **ZERO-SHOT PROMPT**. In the next section you will use other strategies to improve the output.

**Business user:** Facilities coordinator  
**Purpose:** Turn an emailed repair request into a consistent intake record.

**Provided input**

```text
From: Maya Chen
To: Facilities Service Desk
Subject: Loose handrail before Friday tour

The handrail in the east stairwell on floor 3 of Pioneer Hall is loose at the
lower wall bracket. I noticed it at 9:15 a.m. on September 2. No one has been
injured, and the stairwell is still open. Please repair it before the visitor
tour this Friday if possible. I can meet a technician after 1:00 p.m. Call me
at extension 5521.
```


#### TODO - INSTRUCT 🔧

In [8]:
# 🔧 TODO - INSTRUCT
# Strategy: zero-shot — clear directions only, no examples.
# Your GOAL here is to extract the following pieces of information from the email.
# Extract reporter name, location, problem description, date and time observed, urgency or deadline, and contact information."

initial_instruction_extraction = """
Extract the following information from the email below.
If a piece of information is not present in the email, write "Not specified" instead of guessing.

- Reporter name:
- Location:
- Problem description:
- Date and time observed:
- Urgency or deadline:
- Contact information:

Email:
{email_text}

"""

In [9]:
# Run this to generate output after updating your instruction. It will take at least 20 seconds to run each loop.

for i in range(3):

  initial_record_extraction = run_part1(
      task_key="extraction",
      stage="initial",
      instruction=initial_instruction_extraction,
      source=PART1_SOURCES["extraction"],
  )
  display_record(initial_record_extraction)

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T22:47:45+00:00  
**Elapsed:** 132.06 seconds

Here's the extracted information:

- Reporter name: Maya Chen
- Location: Pioneer Hall (east stairwell on floor 3)
- Problem description: Loose handrail
- Date and time observed: September 2, 1999, at 9:15 a.m.
- Urgency or deadline: Urgent - needs to be repaired before the visitor tour this Friday.
- Contact information: Extension 5521

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T22:49:57+00:00  
**Elapsed:** 4.24 seconds

Here is the extracted information from the email:

- Reporter name: Maya Chen
- Location: Pioneer Hall, east stairwell, floor 3
- Problem description: Loose handrail
- Date and time observed: September 2, 9:15 a.m.
- Urgency or deadline: Prioritize repair before the visitor tour Friday
- Contact information: Extension 5521

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T22:50:02+00:00  
**Elapsed:** 4.29 seconds

Here’s the extracted information from the email, presented as requested:

- Reporter name: Maya Chen
- Location: Pioneer Hall, east stairwell on floor 3
- Problem description: Loose handrail
- Date and time observed: 9:15 a.m. on September 2
- Urgency or deadline: Possible repair before the visitor tour this Friday
- Contact information: Extension 5521

### TODO - REFLECT 🖊

Initial-Output Diagnosis and Revision Plan



**Specific weakness in the initial output:**

🖊 A specific weakness of the initial output is that the model may have generated information not present in the original email. In the first run, the date was printed as "September 2, 1999, at 9:15 a.m.," but in the second and third runs, it was printed as "September 2, 9:15 a.m. without the year" and "9:15 a.m. on September 2," respectively. If the original email does not contain the information “1999,” then the year in the first result is information generated by the model. This shows the output deviation between the model’s hallucination and execution.

**Speed of Output:**

🖊 The average CPU execution time could not be determined because the model was not tested in a CPU environment.

🖊 The model took an average of 46.86 seconds to produce the outputs when using a GPU. The first run took considerably longer, likely because of the initial model-loading time (cold start).

**Planned instruction change:**

🖊 The modified prompt will add the instruction: "Do not infer or generate any information not specified in the original email, and if there is no information, write 'Not specified.'" This instruction will help prevent the model from arbitrarily generating information not present in the original text, such as the year, and improve the accuracy and consistency of the output.

**Prompting strategy for the revision:**

🖊 The modified prompt will use a **few-shot prompting** strategy. Provide the model with an example of extracting accurate information and show how to mark information not present in the original text as "Not specified." This strategy is suitable for reducing the output deviation between the random generation of information and execution, because it guides the model to follow the provided output pattern.

### Revise your prompt with a strategy to address the issue you noted above with the model output quality.

### TODO - INSTRUCT 🔧

In [10]:
# 🔧 TODO. Write the revised instruction.
revised_instruction_extraction = """

Extract the following information from the email below.
Only use information that is explicitly stated in the email.
If a piece of information is not present, write "Not specified" — do not guess or infer it.

Example:
Email: "From: John Park. Subject: broken window in Room 210.
The window latch broke this morning. Please fix before end of week."

Extracted:
- Reporter name: John Park
- Location: Room 210
- Problem description: Broken window latch
- Date and time observed: Not specified
- Urgency or deadline: Before end of week
- Contact information: Not specified

Now extract from this email:

- Reporter name:
- Location:
- Problem description:
- Date and time observed:
- Urgency or deadline:
- Contact information:

Email:
{email_text}

"""


In [11]:
# run this to create the output. It will take at least 20 seconds to run each loop.
for i in range(3):

  revised_record_extraction = run_part1(
      task_key="extraction",
      stage="revised",
      instruction=revised_instruction_extraction,
      source=PART1_SOURCES["extraction"],
  )
  display_record(revised_record_extraction)

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:05:52+00:00  
**Elapsed:** 4.11 seconds

- Reporter name: Maya Chen
- Location: Pioneer Hall
- Problem description: Loose handrail
- Date and time observed: September 2 at 9:15 a.m.
- Urgency or deadline: If possible, repair before the visitor tour this Friday.
- Contact information: Extension 5521

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:05:56+00:00  
**Elapsed:** 5.58 seconds

- Reporter name: Maya Chen
- Location: Pioneer Hall, east stairwell on floor 3
- Problem description: Loose handrail
- Date and time observed: September 2 at 9:15 a.m.
- Urgency or deadline: Please repair it before the visitor tour this Friday if possible
- Contact information: Call me at extension 5521

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:06:02+00:00  
**Elapsed:** 4.16 seconds

- Reporter name: Maya Chen
- Location: Pioneer Hall, east stairwell on floor 3
- Problem description: Loose handrail
- Date and time observed: September 2, 9:15 a.m.
- Urgency or deadline: Before the visitor tour this Friday
- Contact information: Extension 5521

### TODO - REFLECT 🖊

Improvement and Human Review

**Effect of the revision:**

🖊The revised prompt substantially improved the output results. In the initial output, the year "1999" was added to the original text on the first run, but after modification, all three subsequent outputs extracted it accurately as "September 2 at 9:15 a.m." without the year. Therefore, the instruction not to generate information not present in the original text, along with few-shot examples, was effective in reducing the model’s hallucinations.

**Role of the prompting strategy:**

🖊The example of John Park and Room 210 used in the few-shot prompting showed the model to extract only the information present in the original text and mark any missing information as "Not specified." This example improved the accuracy and consistency of the output by encouraging the model not to arbitrarily add years not present in the original text even in real emails.

**Human review still required:**

🖊Although the model’s output has been improved, it still requires human review before being used in real-world tasks. For example, the location was described differently as "Pioneer Hall" or "Pioneer Hall, east stairwell on floor 3," and the contact information also differed in format, with "Extension 5521" and "Call me at extension 5521." Therefore, the level of detail of location information and the contact format should be standardized, and a human should make the final verification to ensure that the extracted information matches the original.

**Output Variability**

🖊In the three modified outputs, all key facts such as name, problem, and deadline were identical, and no facts or other information were generated. The differences between implementations were limited to minor formal differences, such as the level of detail in location information and the way contacts are represented. Therefore, all three modified outputs can be used as is, or used effectively after slight formatting adjustments. Compared with the initial output, where only two of the three were faithful to the original text, the reliability and consistency of the result have improved.

## Part 1.2: Summarization

**Business user:** Customer-service supervisor  
**Purpose:** Prepare a concise escalation summary without losing financial details.

**Provided input**

```text
Customer Elena Ruiz reported that order OR-8841 was charged twice. The
original $186.40 charge posted on August 6, and a second $186.40 charge posted
on August 8 after she refreshed the checkout page. The order itself arrived on
August 10 and was correct. Agent Malik opened case CS-2197 on August 11 and
asked Billing to investigate. Billing has not yet confirmed whether the second
entry is a settled charge or a temporary authorization. Elena wants the second
charge removed if it settled, but she does not want the order canceled. She
asked for an update by August 13 because her card payment is due August 14.
```


**Your task:** A customer-service supervisor needs a concise escalation summary to hand off to a billing team. The model should condense the case above without losing any financially important detail — amounts, dates, case numbers, and the customer's stated deadline all matter.

**What to do in the cells below:**
- **First code cell:** Replace the `TODO` with your zero-shot instruction. Specify the audience (the billing team), the required level of detail, and any format constraints (length, structure). Run the cell and leave the output visible.
- **Diagnosis markdown cell:** Compare the output against the source. Identify one specific weakness — for example, a missing amount, a dropped date, or a format that buries the urgency.
- **Second code cell:** Write your revised instruction addressing that weakness. Remember: at least two of your four tasks must apply a named strategy with an explanation.
- **Evaluation markdown cell:** Explain what improved, what the strategy contributed, and which facts in the summary would need human verification before the billing team acts on them.

### TODO - INSTRUCT 🔧

In [12]:
# 🔧 Strategy: zero-shot — clear directions only, no examples.
initial_instruction_summarization = """

You are writing a concise escalation summary for the billing team.
Summarize the case below in 3-5 sentences.

You must preserve the following details exactly as stated:
- The order number
- Both charge amounts and the dates they posted
- The case number
- The unresolved billing question
- The customer's requested deadline

Case:
{case_text}
"""

In [13]:
for i in range(3):

  initial_record_summarization = run_part1(
      task_key="summarization",
      stage="initial",
      instruction=initial_instruction_summarization,
      source=PART1_SOURCES["summarization"],
  )
  display_record(initial_record_summarization)

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:10:34+00:00  
**Elapsed:** 4.52 seconds

Here is a concise escalation summary for the case:

Subject: Billing Issue - Order OR-8841 – Urgent Resolution Needed

Elena Ruiz is contacting us because two unauthorized charge entries were posted to her account, resulting in a double charge for order OR-8841. The charges were initially posted on August 6th and August 8th.  Billing’s investigation is currently ongoing and assessing whether the second charge represents a settled amount or a temporary authorization. Elena requests a resolution, specifically removing the second charge if it's a settled payment and a deadline of August 13th for payment.

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:10:39+00:00  
**Elapsed:** 4.72 seconds

Here’s a 3-5 sentence summary of the case:

Elena Ruiz, Case CS-2197, submitted a complaint regarding twice-charged orders OR-8841. The initial $186.40 charge appeared on August 6th and followed by a $186.40 charge on August 8th, despite the order arrival date being August 10th. Her requesting removal of the second charge, if it’s a settlement, while maintaining the original order, is causing concern. She demands an urgent update by August 13th due to a payment due date.

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:10:44+00:00  
**Elapsed:** 5.39 seconds

Here’s a concise escalation summary for the billing team:

**Subject: Billing Investigation - Case CS-2197**

Customer Elena Ruiz is reporting two duplicate billing charges – a $186.40 charge on August 6 and another $186.40 on August 8 – totaling $372.80. The order arrived on August 10 and is accurate, prompting the billing team to investigate. Agent Malik initiated the issue on August 11, requesting confirmation of whether the second charge is a settled or temporary authorization, and Ms. Ruiz’s August 13 deadline for her payment is creating a critical timeframe.

### TODO - REFLECT 🖊

Initial-Output Diagnosis and Revision Plan

Use Strategy: zero-shot — clear directions only, no examples.

**Specific weakness in the initial output:**

🖊In the three initial runs, the required identifiers (order number, case number) are missing differently each time — run 1 lacks the case number (CS-2197), and runs 2 and 3 lack the order number (OR-8841). There is not a single output containing both of the two required identifiers among the three executions, making it difficult for the billing team to track the case in the system.

**Planned instruction change:**

🖊Since the instruction to "definitely include" was not enough on its own, we plan to revise the instruction to assign the context (role) that omitting ** causes accountability issues to the model, so it pays more attention to missing identifiers.

**Prompting strategy for the revision:**

🖊Apply the Persona strategy. By assigning the role of "billing operations analyst," where traceability is a compliance requirement, to the model, the context within the prompt that identifiers (order number, case number) must not be omitted is reinforced. Since this task failed twice using a simple zero-shot approach, adding role-based context so that the model approaches it with the frame of "this is the official documentation that should be auditable" is well suited to this weakness.

**Your task:** A procurement analyst needs to send the plant manager a status update about a delayed shipment. The draft must be factually accurate, avoid making commitments that have not been approved (e.g., expedited shipping has not been authorized), and convey appropriate urgency without overstating the risk.

**What to do in the cells below:**
- **First code cell:** Replace the `TODO` with your zero-shot instruction. Specify the intended recipient (the plant manager), the tone, and any constraints on what the draft should or should not commit to. Run the cell and leave the output visible.
- **Diagnosis markdown cell:** Review the draft for any facts that differ from the source, commitments the model made that are not supported, or tone and structure problems.
- **Second code cell:** Write your revised instruction. Consider whether a persona strategy (e.g., "You are a procurement analyst...") or chain-of-thought reasoning helps the model avoid unsupported commitments.
- **Evaluation markdown cell:** Explain what improved, what the strategy contributed, and what a human analyst must check before sending the draft.

### Revise your prompt with a strategy to address the issue you noted above with the model output quality.

### TODO - INSTRUCT 🔧

In [14]:
# 🔧 Identify your strategy at the start of the instruction string, e.g.: "Strategy: persona"
revised_instruction_summarization = """

Strategy: persona

You are a billing operations analyst preparing an escalation summary for the billing team.
Every escalation you write must be traceable in the billing system, so you never omit
the order number or case number — omitting either creates a compliance risk.

Summarize the case below in 3-5 sentences for the billing team.
You must include, exactly as stated:
- Order number
- Case number
- Both charge amounts and the dates they posted
- The unresolved billing question
- The customer's requested deadline

Case:
{case_text}

"""

In [15]:
for i in range(3):

  revised_record_summarization = run_part1(
      task_key="summarization",
      stage="revised",
      instruction=revised_instruction_summarization,
      source=PART1_SOURCES["summarization"],
  )
  display_record(revised_record_summarization)

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:13:26+00:00  
**Elapsed:** 4.88 seconds

Here's a summary of the case, tailored for the billing team, adhering to your requirements:

**Order Number:** OR-8841
**Case number:** CS-2197
**Total Charges:** $186.40 and $186.40
**Affected Dates:** August 6th – August 8th
**Resolved Issue:** The customer, Elena Ruiz, has reported two duplicate charge entries for order OR-8841, which originated on August 6th and August 8th. Currently, the billing system is unable to definitively determine whether the second charge is a settled payment or a temporary authorization request, requiring urgent investigation. Elena has requested a resolution – that the second charge be removed if it was a settled transaction, but not to cancel the order.  She’s requesting an update by August 13 to satisfy her payment due date.

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:13:31+00:00  
**Elapsed:** 5.96 seconds

Okay, here's a three-to-five sentence summary of the case, focused on traceability and including the required elements – with the order number and case number exactly as they appear in the provided text:

**Summary:**

Case CS-2197 – Elena Ruiz has filed an escalation regarding two duplicate charges for order OR-8841, which were posted on August 8 and August 6, respectively.  The original charge of $186.40 was associated with August 6, while a subsequent $186.40 charge appeared after a page refresh on August 8. The billing team needs to confirm whether these two charges constitute settled obligations or temporary authorizations to determine the best course of action for customer Elena.  Elena requests an update by August 13 to ensure her card payment is processed and a resolution is available by August 14; consequently, the billing team needs to address this discrepancy and potentially reset the charges. 


---

Let me know if you need any revisions to this!

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:13:37+00:00  
**Elapsed:** 4.89 seconds

Okay, here’s a 3-5-sentence summary of the case for the billing team, adhering precisely to your specifications:

**Case Summary - Order OR-8841**

**Case Number:** CS-2197 – Received case from Elena Ruiz regarding duplicate charges on Order OR-8841. Initial charges posted on August 6th ($186.40) and August 8th ($186.40), with the order arriving on August 10th and verified to be valid. Billing's investigation has not yet clarified whether the second $186.40 charge is a settled, temporary authorization, and the customer’s desired deadline is August 13th.  Therefore, the Billing team urgently needs to confirm if the second charge is a settlement or an unauthorized transaction and initiate corrective action.**

### TODO - REFLECT 🖊

Improvement and Human Review

**Effect of the revision:**

🖊 The revised prompt clearly improved the output result. In the first three runs, either the order number or the case number was missing each time, but in the three outputs after correction, both the order number OR-8841 and the case number CS-2197 were included. Therefore, the persona and the list of explicitly required items were effective in resolving the previous identifier omission issue.

**Role of the prompting strategy:**

🖊 The persona strategy, “You never omit the order number or case number—omitting either creates a compliance risk,” assigned a role to the model for the responsible person who must verify both identifiers. In particular, by linking the need to include identifiers to compliance risk, the model was guided to include both order numbers and case numbers without omission in all three runs.

**Human review still required:**

🖊 Even in the revised output, there are parts that still require human review. In the second iteration, the order of the dates was rearranged and presented, and the phrase "reset the charges," which was not in the original text, was added. This can cause confusion over whether the actual processing method is a refund or a claim cancellation. Therefore, the billing team must verify whether the order of dates and the billing processing method match the original text before sending the message.

## Part 1.3: Drafting

**Business user:** Procurement analyst  
**Purpose:** Draft an internal delay notice that does not make unsupported commitments.

**Provided input**

```text
Supplier Northstar Filtration notified Procurement that shipment NF-771,
containing 12 replacement filters, will arrive August 19 instead of August 14.
The plant currently has approximately four days of filter inventory at normal
usage. Northstar offered expedited shipping for an additional fee, but
Procurement has not approved that option. Operations is checking whether usage
can be reduced safely. The plant manager needs a status update today. No
production shutdown has been scheduled.
```


**Your task:** A service-desk dispatcher needs to route this access request to the correct team and assign the correct urgency level. The routing and urgency definitions are included in the source text above — the model should apply them, not invent its own categories. The model should also flag that required information (employee ID and approval record) is missing.

**What to do in the cells below:**
- **First code cell:** Replace the `TODO` with your zero-shot instruction. Tell the model to output the routing category, urgency level, and any missing information that blocks processing, based strictly on the definitions provided. Run the cell and leave the output visible.
- **Diagnosis markdown cell:** Check whether the model applied the urgency definitions correctly (does the workaround change the urgency?), used only the defined categories, and flagged the missing fields. Identify the most significant gap.
- **Second code cell:** Write your revised instruction. Chain-of-thought is often effective here — prompting the model to reason through each definition before giving a final answer can reduce misclassification.
- **Evaluation markdown cell:** Explain what improved, what the strategy contributed, and what a human dispatcher must still decide before acting on the model's output.

### TODO - INSTRUCT 🔧

In [16]:
# 🔧 TODO - INSTRUCT
# Strategy: zero-shot — clear directions only, no examples.
initial_instruction_drafting = """

You are a procurement analyst. Write a short internal status update for the plant manager
about a delayed filter shipment, based only on the information provided below.

Do not state or imply that expedited shipping has been approved.
Do not state or imply that a production shutdown has been scheduled.
Only include facts that are explicitly stated in the source text.

Source:
{source_text}

"""

In [17]:
# Run this to generate output after updating your instruction. It will take at least 20 seconds to run each loop.

for i in range(3):
  initial_record_drafting = run_part1(
      task_key="drafting",
      stage="initial",
      instruction=initial_instruction_drafting,
      source=PART1_SOURCES["drafting"],
  )
  display_record(initial_record_drafting)

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:24:58+00:00  
**Elapsed:** 5.35 seconds

Okay, here’s the internal status update for the plant manager, based solely on the provided source text:

**INTERNALSTATUSUPDATE**

**Subject:** Filter Shipment Delay – NF-771

**To:** Plant Manager

**From:** [Your Name], Procurement Analyst

Today we have notification that shipment NF-771, containing 12 replacement filters, will arrive August 19th instead of August 14th. 

NorthStar Filtration has notified us that the shipment is coming, but this is happening without expedited shipping. Operations is assessing if current inventory levels can be adequately maintained through available usage. No production shutdown has been scheduled.


Let me know if you’d like me to refine this further.

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:25:04+00:00  
**Elapsed:** 4.39 seconds

Okay, here’s a brief internal status update for the plant manager, based on the provided information, keeping it factual and concise:

**Internal Status Update**

Supplier Northstar Filtration notified us that shipment NF-771, for 12 replacement filters, will now arrive on August 19 instead of August 14. We currently have approximately four days of filter inventory. NorthStar offered expedited shipping, but approval is currently pending. Operations is reviewing usage patterns to determine if reduced operations can safely mitigate the impact.  No production shutdowns are scheduled.

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:25:08+00:00  
**Elapsed:** 4.42 seconds

Okay, here's a short internal status update for the plant manager, based *strictly* on the provided text:

Subject: Filter Shipment Delay – NF-771

Gentlemen,

Northstar Filtration has informed us that shipment NF-771 – containing twelve replacement filters – will now arrive approximately two weeks earlier, August 19th, rather than August 14th. 

Our current inventory levels support operations for approximately four days of usage.

We're reviewing usage patterns to determine if modifications to production will be safely implemented. 


– Procurement Team

### TODO - REFLECT 🖊

Initial-Output Diagnosis and Revision Plan

**Specific weakness in the initial output:**

🖊 The initial output showed a problem of certainty inflation, such as changing “has not approved” to “approval is pending,” thereby expressing the original text’s uncertain state more definitively than it actually is.

**Planned instruction change:**

🖊 We will add a rule to the revised instructions that, when expressing approval status or an undecided state, should retain the certainty level of the original text and not replace “not approved” with “pending,” etc.

**Prompting strategy for the revision:**

🖊 By using a chain-of-thought strategy, the model will be trained to distinguish between confirmed facts and unconfirmed information step-by-step before producing the final answer, thereby reducing the problem of expressing uncertain content in a definitive manner.

### Revise your prompt with a strategy to address the issue you noted above with the model output quality.

### TODO - INSTRUCT 🔧

In [18]:
# 🔧 TODO - INSTRUCT
# Identify your strategy at the start of the instruction string, e.g.: "Strategy: chain-of-thought"
revised_instruction_drafting = """
Strategy: chain-of-thought

You are a procurement analyst writing a short internal status update for the plant manager
about a delayed filter shipment, based only on the source text below.

Before writing the update, work through these steps internally:
1. List only the facts that are explicitly confirmed in the source (e.g., new arrival date, current inventory).
2. List anything in the source that is still undecided, pending, or unapproved (e.g., expedited shipping, usage reduction).
3. Make sure your final update preserves the exact certainty level of each item from step 2 —
   do not describe an unapproved option as "pending approval" or an unconfirmed review as if its outcome is already decided.

Then write the final update. Do not include your step-by-step reasoning in the output —
only the final status update.

Source:
{source_text}

"""

In [19]:
# Run this to generate output after updating your instruction.

revised_record_drafting = run_part1(
    task_key="drafting",
    stage="revised",
    instruction=revised_instruction_drafting,
    source=PART1_SOURCES["drafting"],
)
display_record(revised_record_drafting)

**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:29:06+00:00  
**Elapsed:** 3.91 seconds

August 19th. The shipment NF-771 will arrive, instead of August 14th. Northstar Filtration has provided the new arrival date. Operations is assessing the impact on filter usage and is reviewing options to reduce usage. }}

### TODO - REFLECT 🖊

Improvement and Human Review

**Effect of the revision:**

🖊 This revision reduced the problem of exaggerated certainty in output, but overall it is hard to see it as a substantial improvement. In the revised output, the fact that the expedited shipment was not approved, the absence of a production shutdown plan, and the omission of a four-day inventory buffer were noted. Also, the sentence starts awkwardly with "August 19th" and ends with `}}A format error also occurred with the addition of `. Therefore, instead of reducing the existing problem, a more serious problem arose in terms of completeness and formalism.

**Role of the prompting strategy:**

🖊 The chain-of-thought strategy guided the model to distinguish between fixed facts and unfixed information, but the instructions for the gemma3:1b model may have been overly complex. During the process in which the model followed multiple inference steps, it omitted some key information and failed to maintain the format of the final document stably. Therefore, for this task, a simple zero-shot strategy that explicitly states the required information and disallows it may be more suitable than a chain-of-thought.

**Human review still required:**

🖊 Before sending the documents, the procurement analyst must verify whether the urgent shipment has been approved, whether there is a production shutdown, and whether a four-day inventory buffer is included. It also verifies whether the shipping date and other facts match the original text, and the closing brace `}}After removing unnecessary characters such as `, it should be revised into a proper memo format including the subject, recipients, body, and signature.

# Part 2: Controlled Four-Model Evaluation

In Part 1, you were free to revise your instructions and iterate. Part 2 is different: you write **one** instruction and send it to all four models **unchanged**. This is a controlled experiment — the only variable is the model itself.

**Why controlled?** If you give different prompts to different models, any differences in output could come from your instruction, not from the model. A shared, unmodified prompt isolates the model as the only variable and makes your comparisons valid.



## Service Issue Details

```text
REQUEST SR-2401
Site: North Distribution Center
Reported by: Luis Ortega, extension 4410
At 6:40 a.m. on June 12, the quality-control freezer display read 18 F. Its
required operating range is 0-5 F. Temperature-sensitive calibration
materials were moved to the backup freezer. Staff reset the alarm twice,
but it returned both times. No employee injury was reported.
```


In [21]:
multimodeltext = """
REQUEST SR-2401
Site: North Distribution Center
Reported by: Luis Ortega, extension 4410
At 6:40 a.m. on June 12, the quality-control freezer display read 18 F. Its
required operating range is 0-5 F. Temperature-sensitive calibration
materials were moved to the backup freezer. Staff reset the alarm twice,
but it returned both times. No employee injury was reported.
"""

### TODO - INSTRUCT 🔧

In [22]:
# 🔧 TODO - INSTRUCT
# Strategy: zero-shot — clear directions only, no examples.
initial_instruction_extraction = """

Extract the following information from the email below.
Only use information that is explicitly stated in the email.
If a piece of information is not present, write "Not specified" — do not guess or infer it.

Example:
Email: "From: John Park. Subject: broken window in Room 210.
The window latch broke this morning. Please fix before end of week."

Extracted:
- Reporter name: John Park
- Location: Room 210
- Problem description: Broken window latch
- Date and time observed: Not specified
- Urgency or deadline: Before end of week
- Contact information: Not specified

Now extract from this email:

- Reporter name:
- Location:
- Problem description:
- Date and time observed:
- Urgency or deadline:
- Contact information:

Email:
{email_text}

"""



In [23]:
# Run this after revising the

comparison_prompt = compose_prompt(
    initial_instruction_extraction,
    multimodeltext,
)

comparison_records = {}
for model in REQUIRED_MODELS:
    print(f"Running independent comparison: {model}")
    comparison_records[model] = chat_once(
        model,
        comparison_prompt,
        f"comparison_{model}",
    )
    display_record(comparison_records[model])


Running independent comparison: gemma3:1b


**Model:** `gemma3:1b`  
**Recorded:** 2026-09-07T23:37:29+00:00  
**Elapsed:** 5.36 seconds

Okay, here's the information extracted from the email, based solely on what is explicitly stated:

- Reporter name: Luis Ortega
- Location: North Distribution Center
- Problem description: The quality-control freezer display read 18 F and the temperature-sensitive calibration materials were moved to the backup freezer.
- Date and time observed: 6:40 a.m. on June 12
- Urgency or deadline: Not specified
- Contact information: Not specified

Running independent comparison: gemma3:4b


**Model:** `gemma3:4b`  
**Recorded:** 2026-09-07T23:37:34+00:00  
**Elapsed:** 40.77 seconds

- Reporter name: Luis Ortega
- Location: North Distribution Center
- Problem description: Quality-control freezer display read 18 F, outside of the required operating range (0-5 F). Alarm repeatedly returned after reset.
- Date and time observed: June 12 at 6:40 a.m.
- Urgency or deadline: Not specified
- Contact information: Not specified

Running independent comparison: llama3.2:1b


**Model:** `llama3.2:1b`  
**Recorded:** 2026-09-07T23:38:15+00:00  
**Elapsed:** 13.09 seconds

Here is the extracted information:

- Reporter name: Luis Ortega
- Location: North Distribution Center
- Problem description: Temperature-sensitive calibration materials moved to backup freezer
- Date and time observed: Not specified
- Urgency or deadline: 
- Contact information: Not specified

Running independent comparison: llama3.2:3b


**Model:** `llama3.2:3b`  
**Recorded:** 2026-09-07T23:38:28+00:00  
**Elapsed:** 22.53 seconds

Here are the extracted information from the email:

- Reporter name: Luis Ortega
- Location: North Distribution Center
- Problem description: Quality-control freezer display read 18 F (outside required operating range)
- Date and time observed: June 12, 6:40 a.m.
- Urgency or deadline: Not specified
- Contact information: extension 4410

### TODO - REFLECT 🖊

Model Evaluation

**How similar or different were the outputs?:**

🖊 All four models had the name and location extracted identically, but differences appeared in the problem description and contact information. In particular, llama3.2:1b omitted the key issue that the freezer temperature rose to 18°F and only mentioned material movement, and the only model that accurately extracted extension 4410 from the original text was llama3.2:3b.


**How long did each model take to run?:**

🖊 The execution times were gemma3:1b:5.36 s, gemma3:4b:40.77 s, llama3.2:1b:13.09 s, and llama3.2:3b:22.53 s. Within the same model family, models with larger parameters were slower, but between different families, execution speed could not be predicted based solely on model size.

**Which model was "Best"?:**

🖊 For this source and prompt, llama3.2:3b was the most suitable. This model specifically describes temperature anomalies and temperature ranges, uniquely extracts the contact extension 4410 accurately, and runs faster than gemma3:4b. However, this judgment is based on the results of a single run of this task, so it does not mean that llama3.2:3b is always superior in every task.

# Part 3: Reflect on the Assignment



#### TODO - FINAL REFLECTION 🖊

Write a **150–250 word** reflection and answer the

1. **Prompt effect (Part 1):** Which revision across your four Part 1 tasks produced the largest improvement? What specifically changed in the model's output, and why did the revised instruction work better?

🖊The summarization revision produced the largest improvement. All three initial runs omitted either the order number or case number; after adding a persona framing identifiers as a compliance requirement, all three revised runs included both. The explicit accountability framing, not just a longer instruction, drove the change.

2. **Strategy fit:** For each named strategy you applied (few-shot, chain-of-thought, or persona), explain why you chose it for that task. If you kept zero-shot for any revision, explain why plain directions were sufficient.

🖊 Few-shot suited extraction because the example demonstrated marking missing facts as "Not specified," which stopped a fabricated year from appearing. Persona suited summarization by framing the output as an auditable document, reinforcing identifier retention. Chain-of-thought was applied to drafting but underperformed — the added reasoning steps caused gemma3:1b to drop key facts and produce malformed output, showing plain zero-shot directions can outperform more complex strategies on small models.

3. **Model differences (Part 2):** Where did the four models diverge most noticeably — in accuracy, completeness, format adherence, or something else? Give a specific example.

🖊 The models diverged most on completeness: llama3.2:1b omitted the core temperature anomaly (18°F, outside 0–5°F range), while only llama3.2:3b correctly extracted the contact extension (4410).

4. **Size versus quality:** Did the larger model in each family (Gemma 3 4B, Llama 3.2 3B) consistently produce better outputs than the smaller one? Were there cases where the size difference did not predict quality?

🖊 Larger models weren't consistently better. Within families, 4b/3b captured more detail than 1b, but across families llama3.2:1b (13.09s) still missed the central fact that gemma3:1b (5.36s) caught — size alone didn't predict quality.

5. **Runtime trade-offs:** How did elapsed time vary across the four models? Was the quality gain from the slower models worth the additional time, given the nature of this task?

🖊 Elapsed times ranged from 5.36s (gemma3:1b) to 40.77s (gemma3:4b). Gemma3:4b's extra time wasn't justified — llama3.2:3b (22.53s) delivered comparable or better completeness faster.

6. **Business risk:** Identify one specific output — from either Part 1 or Part 2 — that would cause a real problem if a person acted on it without review. What is the specific risk, and what kind of human check would catch it?

🖊 The chain-of-thought drafting output posed the clearest risk: it dropped the fact that expedited shipping was unapproved and that no shutdown was scheduled. Sending it unreviewed could mislead the plant manager into believing shipping was already expedited. A human must compare the draft against the source before sending.

7. **Generalization:** Based on what you observed, under what conditions would a small local model like these be a reasonable choice for a business task? Under what conditions would you want a larger or cloud-hosted model instead?

🖊 Small local models are reasonable for low-stakes, human-reviewed internal tasks with simple, well-defined outputs. A larger or cloud-hosted model is preferable when completeness, nuanced judgment, or external/compliance-facing accuracy matters more than speed or cost.